### ISSUE TITLE
Suggest complex objects in a trial?

### ISSUE THREAD

**Issue by matthiasschuurmans** [optuna/optuna#2041]
Dear Optuna creators, thanks a lot for this package, it has helped me tremendously for a while now already. I especially like the nice visualization of a finished study object!

My question is: is it somehow possible to suggest complex objects in a trial? In https://optuna.readthedocs.io/en/stable/tutorial/002_configurations.html it shows that you can suggest strings for categorical variables, for example an optimizer for a keras network, but each optimizer has its own parameters that are worth tuning and suggesting strings is only possible due to keras having such aliases. Of course I could recreate the object from its individually suggested values like how it happens in the linked documentation, but it would be very handy to be able to immediately suggest the object.

I.e for an optimizer, I'm hoping for something like this:

```
trial.suggest_categorical('optimizer', [
    keras.optimizers.Adam(
        learning_rate=trial.suggest_float('learning_rate', 0.0001, 0.001)
    ),
    keras.optimizers.SGD(
        learning_rate=trial.suggest_float('learning_rate', 0.0001, 0.001),
        momentum=trial.suggest_float('momentum', 0.1, 0.9)
    )
])
```

This would allow optimizing the optimizer (and any other objects that have their own parameters to optimize - `kernel_regularizer` with changing `l1` and/or `l2` values for example) in one go, as opposed to having a bunch of code to recreate the optimizer with its suggested individual parameters afterwards. But if I try this atm, I get

```
Choices for a categorical distribution should be a tuple of None, bool, int, float and str for persistent storage but contains <tensorflow.python.keras.optimizer_v2.adam.Adam object at 0x0000029C8F695208> which is of type Adam.
[W 2020-11-20 09:17:44,060] Trial 1 failed because of the following error: ValueError('CategoricalDistribution does not support dynamic value space.')
```

Thanks for any help!

**Comment by hvy** [comment]
I think that we theoretically could support more general types of use cases such as the one you provide. That would require some changes to the internals though since the list of choices must be persisted in the storage so they must be serializable. We in other words cannot allow arbitrary objects. I'm not entirely sure if that it would simplify the lives of users or developers however since it make the API more relaxed and prone to bugs (e.g. having stateful objects among the choices). 

This is more of a questions on the optimization performance. You're suggesting both the "learning_rate" and the "momentum" but presumably only using either. This might be suboptimal when doing optimization with relative samplers (such as the CMA-ES) which would try to do joint sampling of all parameters assuming that all parameters played a role in the optimizatin.

**Comment by matthiasschuurmans** [comment]
Hi @hvy , thanks for replying. I can imagine it would make some lives more difficult, so maybe it's a bad idea to do a lot of work for, especially if there are not a lot of people that would want to have such a change. And I do have a workaround as your documentation suggests.

As for joint sampling and optimization, I don't think I understand you, but that might be because I haven't worked with SGD optimizer yet, it was just an example of how I imagine I would want to optimize the optimizers and their parameters in one go. How am I "presumably only using either"? 

As you say, I imagine that in most optimization situations with more than one parameter, the parameters being optimized have a relation to each other, and Optuna usually manages to find a good value for each such that they are pretty good together (i.e num leaves, min data in leaf, max depth and number of trees in LightGBM). How is that different from the SGD optimizer needing to have both a tuned momentum and learning rate?

**Comment by hvy** [comment]
> How am I "presumably only using either"?

Sorry, I was probably not very clear. In the code snippet above, you'd be suggesting `momentum`, even when `optimizer` would be Adam. It may work but might be suboptimal. The reason is that the underlying sampler will think that the search space is flat although it in practice is conditioned (on `optimizer`), and that the value of `momentum` affects the objective value, even you'd be using Adam and `momentum` doesn't matter.

**Comment by matthiasschuurmans** [comment]
Aah right, makes sense. Thanks again for the answers, I'll close my issue.
